### PR TITLE
Packet bare metal cloud hosting platform provider

### PR DESCRIPTION
**Pull request by crunchywelch**
First effort on this, let me know if you'd like to have tests written or if this is a good place to start with the initial PR. Thanks!

### CODE REVIEW

File: builtin/providers/packet/config.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package packet
+
+import (
+	"github.com/packethost/packngo"
+)
+
+const (
+	consumerToken = "aZ9GmqHTPtxevvFq9SK3Pi2yr9YCbRzduCSXF2SNem5sjB91mDq7Th3ZwTtRqMWZ"
```
Comment (pmoust) [inline builtin/providers/packet/config.go]: What's this?

File: builtin/providers/packet/config.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+type Config struct {
+	AuthToken string
+}
+
+// Client() returns a new client for accessing packet.
+func (c *Config) Client() *packngo.Client {
+	return packngo.NewClient(consumerToken, c.AuthToken)
```
Comment (pmoust) [inline builtin/providers/packet/config.go]: Same, https://www.packet.net/dev/api/#authentication

Is the `consumerToken` something undocumented?

File: builtin/providers/packet/config.go (hunk 1 above)
Comment (crunchywelch) [inline builtin/providers/packet/config.go]: it is used by us to identify the source of a request, this is the one we issued for terraform. basically, it is not required, but if it is not present then we rate limit the requests. it is also not used for authentication, that is what the end user's auth token is used for. does that make sense?

File: builtin/providers/packet/config.go (hunk 2 above)
Comment (crunchywelch) [inline builtin/providers/packet/config.go]: I made a note on your other note, happy to add to the docs if that will be clear to end users, but it is not a parameter that should / will change unless folks are editing source code for some reason, like forking terraform for another service or something. does that make sense?

**Comment by crunchywelch** [comment]
Also, I have a test for the device resource, but it requires a project id, and I'm not sure the best way to inject that into the config variables. If someone can point me to an example I'll drop that in as well, otherwise will take a deeper look later this week.

**Comment by rimusz** [comment]
@radeksimko would be nice to get it merged to master, so it can be released with the next terraform version.
Packet.net offers very nice bare-metal servers service, and to have it's support in terraform would make much easier to deploy it's servers.

**Comment by josephholsten** [comment]
@phinze looks ready for :eyes:, has tests & doc too!

File: builtin/providers/packet/resource_packet_device.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+	if err != nil {
+		return fmt.Errorf("Error creating device: %s", err)
+	}
+
+	// Assign the device id
+	d.SetId(newDevice.ID)
+
+	log.Printf("[INFO] Device ID: %s", d.Id())
+
+	_, err = WaitForDeviceAttribute(d, "active", []string{"provisioning"}, "state", meta)
```
Comment (betawaffle) [inline builtin/providers/packet/resource_packet_device.go]: @crunchywelch Should you perhaps add `"queued"` to the pending states?

**Comment by phinze** [comment]
This is looking good! I'll pull it in as-is and improvements can be made on separate PRs.

> Also, I have a test for the device resource, but it requires a project id, and I'm not sure the best way to inject that into the config variables. If someone can point me to an example I'll drop that in as well, otherwise will take a deeper look later this week.

@crunchywelch generally for tests that require special settings like that we just have test-specific environment variables and skip the test with a message if they're not set.

**Pull request merged**